alf/environments/igibson/igibson_tasks.py

```diff
@@ -0,0 +1,816 @@
+        """Sample robot initial pose and target position.
+
+        Args:
+            env (igibson.igibson_env.iGibsonCustomEnv): environment instance
+        Returns:
+            initial and target pose information:
+            - initial_pos (numpy.ndarray): initial position
+            - initial_orn (numpy.ndarray): initial orientation
+            - target_pos (numpy.ndarray): target position
+        """
+        _, initial_pos = env.scene.get_random_point(floor=self.floor_num)
+        max_trials = 100
+        dist = 0.0
+        for _ in range(max_trials):
+            _, target_pos = env.scene.get_random_point(floor=self.floor_num)
+            if env.scene.build_graph:
+                _, dist = env.scene.get_shortest_path(
+                    self.floor_num,
+                    initial_pos[:2],
+                    target_pos[:2],
+                    entire_path=False)
+            else:
+                dist = l2_distance(initial_pos, target_pos)
+            if self.target_dist_min < dist < self.target_dist_max:
+                break
+        if not (self.target_dist_min < dist < self.target_dist_max):
+            print("WARNING: Failed to sample initial and target positions")
+        initial_orn = np.array([0, 0, np.random.uniform(0, np.pi * 2)])
+        return initial_pos, initial_orn, target_pos
+
+    def reset_scene(self, env):
+        """Task-specific scene reset: get a random floor number first.
+
+        Args:
+            env (igibson.igibson_env.iGibsonCustomEnv): environment instance
+        """
+        self.floor_num = env.scene.get_random_floor()
+        super(VisualPointNavRandomTask, self).reset_scene(env)
+
+    def reset_agent(self, env):
+        """Reset robot initial pose.
+
+        Sample initial pose and target position, check validity, and land it.
+
+        Args:
+            env (igibson.igibson_env.iGibsonCustomEnv): environment instance
+        """
+        reset_success = False
+        max_trials = 100
+
+        # cache pybullet state
+        state_id = p.saveState()
+        for i in range(max_trials):
+            initial_pos, initial_orn, target_pos = \
+                self.sample_initial_pose_and_target_pos(env)
+            reset_success = env.test_valid_position(
+                env.robots[0], initial_pos, initial_orn) and \
+                env.test_valid_position(
+                    env.robots[0], target_pos)
+            p.restoreState(state_id)
+            if reset_success:
+                break
+
+        if not reset_success:
+            logging.warning("WARNING: Failed to reset robot without collision")
+
+        p.removeState(state_id)
+
+        x, y, z = target_pos
+        self.target_pos = np.array([x, y, z + self.z_offset])
+        self.initial_pos = initial_pos
+        self.initial_orn = initial_orn
+
+        super(VisualPointNavRandomTask, self).reset_agent(env)
+
+
+class VisualObjectNavTask(BaseTask):
```

Comment:
This class can be merged with VisualObjectNavFixedTask and VisualPointNavRandomTask. All three tasks can be formulated as randomizing (or fixing) a set of locations (number>=1) when resetting and asking the robot to reach the target locomotion. In the case where there is no object concept, the object id can be an arbitrary integer (e.g., 0).